File: shoring/blackjackEngine/src/main/java/com/panopset/blackjackEngine/HandPlayer.java
package com.panopset.blackjackEngine;

import static com.panopset.blackjackEngine.CommandDefinitions.*;
import com.panopset.compat.Logop;

public class HandPlayer extends Hand {

  private boolean isSplit = false;
  private final BlackjackConfiguration c;

  private Wager wager;
  private boolean surrendered = false;
  private boolean selectedEvenMoney = false;
  private String action;

  public HandPlayer(HandPlayer handPlayer) {
    super();
    c = handPlayer.c;
    isSplit = handPlayer.isSplit;
    finalized = handPlayer.isFinal();
    wager = new Wager(handPlayer.wager);
    surrendered = handPlayer.surrendered;
    selectedEvenMoney = handPlayer.selectedEvenMoney;
    for (BlackjackCard card : handPlayer.cards) {
      getCards().add(card);
    }
    setMessage(handPlayer.getMessage());
    setAction(handPlayer.getAction());
  }

  public HandPlayer(final BlackjackGameEngine bge) {
    super();
    c = bge.getConfig();
    wager = new Wager(bge.getBankroll());
  }

  @Override
  public boolean isNatural21() {
    if (isSplit()) {
      return false;
    }
    return super.isNatural21();
  }

  public boolean isSplit() {
    return isSplit;
  }

  public void setSplit() {
    isSplit = true;
  }

  public Wager getWager() {
    return wager;
  }

  private String msg;

  public void setMessage(String message) {
    msg = message;
  }
  

  @Override
  public void dealCard(final BlackjackCard card) {
    super.dealCard(card);
    setMessage("");
  }

  public String getMessage() {
    if (msg == null) {
      msg = "Please select deal (L)";
    }
    return msg;
  }

  public boolean canDouble() {
    if (!isInitialDeal()) {
      return false;
    }
    if (isSplit()) {
      return c.isDoubleAfterSplitAllowed();
    } else {
      return true;
    }
  }

  public boolean isSurrendered() {
    return surrendered;
  }

  public void surrender() {
    surrendered = true;
    stand();
  }

  public boolean isSelectedEvenMoney() {
    return selectedEvenMoney;
  }

  @Override
  public boolean isDone() {
    if (super.isDone()) {
      return true;
    }
    return isSurrendered();
  }

  public void standWithEvenMoney() {
    stand();
    selectedEvenMoney = true;
  }

  public boolean isCardFacesSplittable() {
    return isCardFacesSplittable(false);
  }

  public boolean isCardFacesSplittableIncludeMessage() {
    return isCardFacesSplittable(true);
  }

  private boolean isCardFacesSplittable(boolean includeMessage) {
    if (cards.size() != 2) {
      if (includeMessage) {
        Logop.warn("Can't split a hit hand");
      }
      return false;
    } else if (!isFaceMatch()) {
      if (includeMessage) {
        Logop.warn("Can't split cards that don't have the same face");
      }
      return false;
    }
    return true;
  }

  private boolean isFaceMatch() {
    return cards.get(0).getCard().getFace() == cards.get(1).getCard().getFace();
  }

  public boolean isDoubleDowned() {
    return getWager().isDoubledDown();
  }
  
  public void setAction(String value) {
    action = value;
  }
  
  public String getAction() {
    if (action == null) {
      action = CMD_DEAL;
    }
    return action;
  }
}

File: code/shoring/blackjackEngine/src/main/kotlin/com/panopset/blackjackEngine/HandPlayer.kt
package com.panopset.blackjackEngine

import com.panopset.compat.Logz

class HandPlayer(val wager: Wager) {
    private val hand = Hand()
    var isSplit = false
    var isSurrendered = false
    var isSelectedEvenMoney = false
    var action: String = CMD_DEAL

    constructor(handPlayer: HandPlayer): this(Wager(handPlayer.wager)) {
        isSplit = handPlayer.isSplit
        isSurrendered = handPlayer.isSurrendered
        isSelectedEvenMoney = handPlayer.isSelectedEvenMoney
        hand.isFinal = handPlayer.hand.isFinal
        action = handPlayer.action
        message = handPlayer.message
        val iterator = handPlayer.hand.getIterator()
        val cardsToAddToHand = ArrayList<BlackjackCard>()
        while (iterator.hasNext()) {
            val card = iterator.next()
            cardsToAddToHand.add(card)
        }
        for (card in cardsToAddToHand) {
            dealCard(card)
        }
    }

    fun removeSecondCard(): BlackjackCard {
        return hand.cards.removeAt(1)
    }

    fun isNatural21(): Boolean {
        return if (isSplit) {
            false
        } else hand.isNatural21()
    }

    fun setSplit() {
        isSplit = true
    }

    var message: String = promptDeal
    fun dealCard(card: BlackjackCard) {
        hand.dealCard(card)
        message = ""
    }

//    var message: String
//        get() {
//            if (msg.isEmpty()) {
//                msg = "Please select deal (L)"
//            }
//            return msg
//        }
//        set(message) {
//            msg = message
//        }

    fun canDouble(configIsDoubleAfterSplitAllowed: Boolean): Boolean {
        if (!hand.isInitialDeal) {
            return false
        }
        return if (isSplit) {
            configIsDoubleAfterSplitAllowed
        } else {
            true
        }
    }

    fun surrender() {
        isSurrendered = true
        hand.stand()
    }

    fun isBustedOrSurrenderred(): Boolean {
        return if (hand.isBusted()) {
            true
        } else {
            isSurrendered
        }
    }

    fun standWithEvenMoney() {
        hand.stand()
        isSelectedEvenMoney = true
    }

    val isCardFacesSplittable: Boolean
        get() = isCardFacesSplittable(false)
    val isCardFacesSplittableIncludeMessage: Boolean
        get() = isCardFacesSplittable(true)

    private fun isCardFacesSplittable(includeMessage: Boolean): Boolean {
        if (hand.cards.size != 2) {
            if (includeMessage) {
                Logz.warn("Can't split a hit hand")
            }
            return false
        } else if (!isFaceMatch) {
            if (includeMessage) {
                Logz.warn("Can't split cards that don't have the same face")
            }
            return false
        }
        return true
    }

    fun isBusted(): Boolean {
        return hand.isBusted()
    }

    fun stand() {
        hand.stand()
    }

    fun getFirstCard(): BlackjackCard {
        return hand.cards[0]
    }

    fun getSecondCard(): BlackjackCard {
        return hand.cards[1]
    }

    fun getHandValue(): Int {
        return hand.value
    }

    fun isInitialDeal(): Boolean {
        return hand.isInitialDeal
    }

    fun hasCards(): Boolean {
        return hand.hasCards()
    }

    fun getBlackjackCards(): List<BlackjackCard> {
        return hand.cards
    }

    fun isFinal(): Boolean {
        return hand.isFinal
    }

    fun isSoft(): Boolean {
        return hand.isSoft
    }

    fun isDone(): Boolean {
        return hand.isDone()
    }

    private val isFaceMatch: Boolean
        get() = hand.cards[0].card.face == hand.cards[1].card.face
    val isDoubleDowned: Boolean
        get() = wager.isDoubledDown

//    init {
//        c = handPlayer.c
//        isSplit = handPlayer.isSplit
//        isFinal = handPlayer.isFinal
//        wager = Wager(handPlayer.wager)
//        isSurrendered = handPlayer.isSurrendered
//        isSelectedEvenMoney = handPlayer.isSelectedEvenMoney
//        for (card in handPlayer.cards) {
//            cards.add(card)
//        }
//        message = handPlayer.message
//        action = handPlayer.action
//    }
}
